// @loaders.gl/wms/dist/es5/wip/wcs-capabilities-loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._typecheckWFSCapabilitiesLoader = exports.WCSCapabilitiesLoader = void 0;
var _parseWcsCapabilities = require("./lib/wcs/parse-wcs-capabilities");
const VERSION = typeof "3.4.15" !== 'undefined' ? "3.4.15" : 'latest';
const WCSCapabilitiesLoader = {
  id: 'wcs-capabilities',
  name: 'WFS Capabilities',
  module: 'wms',
  version: VERSION,
  worker: false,
  extensions: ['xml'],
  mimeTypes: ['application/vnd.ogc.wcs_xml', 'application/xml', 'text/xml'],
  testText: testXMLFile,
  options: {
    wms: {}
  },
  parse: async (arrayBuffer, options) => (0, _parseWcsCapabilities.parseWCSCapabilities)(new TextDecoder().decode(arrayBuffer), options),
  parseTextSync: (text, options) => (0, _parseWcsCapabilities.parseWCSCapabilities)(text, options)
};
exports.WCSCapabilitiesLoader = WCSCapabilitiesLoader;
function testXMLFile(text) {
  return text.startsWith('<?xml');
}
const _typecheckWFSCapabilitiesLoader = WCSCapabilitiesLoader;
exports._typecheckWFSCapabilitiesLoader = _typecheckWFSCapabilitiesLoader;
