// @loaders.gl/wms/dist/es5/wip/wfs-capabilities-loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._typecheckWFSCapabilitiesLoader = exports.WFSCapabilitiesLoader = void 0;
var _parseWfsCapabilities = require("./lib/wfs/parse-wfs-capabilities");
const VERSION = typeof "3.4.15" !== 'undefined' ? "3.4.15" : 'latest';
const WFSCapabilitiesLoader = {
  id: 'wfs-capabilities',
  name: 'WFS Capabilities',
  module: 'wms',
  version: VERSION,
  worker: false,
  extensions: ['xml'],
  mimeTypes: ['application/vnd.ogc.wfs_xml', 'application/xml', 'text/xml'],
  testText: testXMLFile,
  options: {
    wms: {}
  },
  parse: async (arrayBuffer, options) => (0, _parseWfsCapabilities.parseWFSCapabilities)(new TextDecoder().decode(arrayBuffer), options),
  parseTextSync: (text, options) => (0, _parseWfsCapabilities.parseWFSCapabilities)(text, options)
};
exports.WFSCapabilitiesLoader = WFSCapabilitiesLoader;
function testXMLFile(text) {
  return text.startsWith('<?xml');
}
const _typecheckWFSCapabilitiesLoader = WFSCapabilitiesLoader;
exports._typecheckWFSCapabilitiesLoader = _typecheckWFSCapabilitiesLoader;
