// @loaders.gl/wms/dist/es5/wip/wmts-capabilities-loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._typecheckWMTSCapabilitiesLoader = exports.WMTSCapabilitiesLoader = void 0;
var _parseWmtsCapabilities = require("./lib/wmts/parse-wmts-capabilities");
const VERSION = typeof "3.4.15" !== 'undefined' ? "3.4.15" : 'latest';
const WMTSCapabilitiesLoader = {
  id: 'wmts-capabilities',
  name: 'WMTS Capabilities',
  module: 'wms',
  version: VERSION,
  worker: false,
  extensions: ['xml'],
  mimeTypes: ['application/vnd.ogc.wmts_xml', 'application/xml', 'text/xml'],
  testText: testXMLFile,
  options: {
    wms: {}
  },
  parse: async (arrayBuffer, options) => (0, _parseWmtsCapabilities.parseWMTSCapabilities)(new TextDecoder().decode(arrayBuffer), options),
  parseTextSync: (text, options) => (0, _parseWmtsCapabilities.parseWMTSCapabilities)(text, options)
};
exports.WMTSCapabilitiesLoader = WMTSCapabilitiesLoader;
function testXMLFile(text) {
  return text.startsWith('<?xml');
}
const _typecheckWMTSCapabilitiesLoader = WMTSCapabilitiesLoader;
exports._typecheckWMTSCapabilitiesLoader = _typecheckWMTSCapabilitiesLoader;
