// @loaders.gl/wms/dist/es5/lib/services/arcgis/arcgis-image-service.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ArcGISImageServer = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _imageSource = require("../../sources/image-source");
class ArcGISImageServer extends _imageSource.ImageSource {
  constructor(props) {
    super(props);
  }
  async getMetadata() {
    return await this.metadata();
  }
  async getImage(parameters) {
    throw new Error('not implemented');
  }
  async metadata() {
    throw new Error('not implemented');
  }
  exportImage(options) {
    throw new Error('not implemented');
  }
  metadataURL(options) {
    return "".concat(this.props.url, "?f=pjson");
  }
  exportImageURL(options) {
    const bbox = "bbox=".concat(options.bbox[0], ",").concat(options.bbox[1], ",").concat(options.bbox[2], ",").concat(options.bbox[3]);
    const size = "size=".concat(options.width, ",").concat(options.height);
    const arcgisOptions = {
      ...options,
      bbox,
      size
    };
    delete arcgisOptions.width;
    delete arcgisOptions.height;
    return this.getUrl('exportImage', arcgisOptions);
  }
  getUrl(path, options, extra) {
    let url = "".concat(this.props.url, "/").concat(path);
    let first = true;
    for (const [key, value] of Object.entries(options)) {
      url += first ? '?' : '&';
      first = false;
      if (Array.isArray(value)) {
        url += "".concat(key.toUpperCase(), "=").concat(value.join(','));
      } else {
        url += "".concat(key.toUpperCase(), "=").concat(value ? String(value) : '');
      }
    }
    return url;
  }
  async checkResponse(response) {
    if (!response.ok) {
      throw new Error('error');
    }
  }
}
exports.ArcGISImageServer = ArcGISImageServer;
(0, _defineProperty2.default)(ArcGISImageServer, "type", 'arcgis-image-server');
(0, _defineProperty2.default)(ArcGISImageServer, "testURL", url => url.toLowerCase().includes('ImageServer'));
