// @loaders.gl/wms/dist/es5/lib/sources/data-source.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DataSource = void 0;
exports.getFetchFunction = getFetchFunction;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
class DataSource {
  constructor(props) {
    (0, _defineProperty2.default)(this, "fetch", void 0);
    (0, _defineProperty2.default)(this, "loadOptions", void 0);
    (0, _defineProperty2.default)(this, "_needsRefresh", true);
    (0, _defineProperty2.default)(this, "props", void 0);
    this.props = {
      ...props
    };
    this.loadOptions = {
      ...props.loadOptions
    };
    this.fetch = getFetchFunction(this.loadOptions);
  }
  setProps(props) {
    this.props = Object.assign(this.props, props);
    this.setNeedsRefresh();
  }
  setNeedsRefresh() {
    this._needsRefresh = true;
  }
  getNeedsRefresh() {
    let clear = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
    const needsRefresh = this._needsRefresh;
    if (clear) {
      this._needsRefresh = false;
    }
    return needsRefresh;
  }
}
exports.DataSource = DataSource;
function getFetchFunction(options) {
  const fetchFunction = options === null || options === void 0 ? void 0 : options.fetch;
  if (fetchFunction && typeof fetchFunction === 'function') {
    return (url, fetchOptions) => fetchFunction(url, fetchOptions);
  }
  const fetchOptions = options === null || options === void 0 ? void 0 : options.fetch;
  if (fetchOptions && typeof fetchOptions !== 'function') {
    return url => fetch(url, fetchOptions);
  }
  return url => fetch(url);
}
