// @loaders.gl/wms/dist/es5/csw-capabilities-loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._typecheckCSWCapabilitiesLoader = exports.CSWCapabilitiesLoader = void 0;
var _parseCswCapabilities = require("./lib/parsers/csw/parse-csw-capabilities");
const VERSION = typeof "3.4.15" !== 'undefined' ? "3.4.15" : 'latest';
const CSWCapabilitiesLoader = {
  id: 'csw-capabilities',
  name: 'CSW Capabilities',
  module: 'wms',
  version: VERSION,
  worker: false,
  extensions: ['xml'],
  mimeTypes: ['application/vnd.ogc.csw_xml', 'application/xml', 'text/xml'],
  testText: testXMLFile,
  options: {
    csw: {}
  },
  parse: async (arrayBuffer, options) => (0, _parseCswCapabilities.parseCSWCapabilities)(new TextDecoder().decode(arrayBuffer), options),
  parseTextSync: (text, options) => (0, _parseCswCapabilities.parseCSWCapabilities)(text, options)
};
exports.CSWCapabilitiesLoader = CSWCapabilitiesLoader;
function testXMLFile(text) {
  return text.startsWith('<?xml');
}
const _typecheckCSWCapabilitiesLoader = CSWCapabilitiesLoader;
exports._typecheckCSWCapabilitiesLoader = _typecheckCSWCapabilitiesLoader;
