// @loaders.gl/wms/dist/es5/csw-domain-loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._typecheckCSWDomainLoader = exports.CSWDomainLoader = void 0;
var _parseCswDomain = require("./lib/parsers/csw/parse-csw-domain");
const VERSION = typeof "3.4.15" !== 'undefined' ? "3.4.15" : 'latest';
const CSWDomainLoader = {
  id: 'csw-domain',
  name: 'CSW Domain',
  module: 'wms',
  version: VERSION,
  worker: false,
  extensions: ['xml'],
  mimeTypes: ['application/vnd.ogc.csw_xml', 'application/xml', 'text/xml'],
  testText: testXMLFile,
  options: {
    csw: {}
  },
  parse: async (arrayBuffer, options) => (0, _parseCswDomain.parseCSWDomain)(new TextDecoder().decode(arrayBuffer), options),
  parseTextSync: (text, options) => (0, _parseCswDomain.parseCSWDomain)(text, options)
};
exports.CSWDomainLoader = CSWDomainLoader;
function testXMLFile(text) {
  return text.startsWith('<?xml');
}
const _typecheckCSWDomainLoader = CSWDomainLoader;
exports._typecheckCSWDomainLoader = _typecheckCSWDomainLoader;
